### PR TITLE
[v1] Support allowed_token_ids in v1 Sampler

### PR DESCRIPTION
**Pull request by houseroad**
Follow the implementation in vllm/entrypoints/openai/logits_processors.py.

The idea is straightforward, adding a [batch_size x vocab_size] mask tensor, and leverage a list of bools to determine whether to do the inplace masked fill.

- [x] add test
- [x] move some verification in _get_allowed_token_ids_logits_processor to SamplingParam.

Test with 
- pytest tests/v1/sample/test_sampler.py
- pytest tests/v1/worker/test_gpu_input_batch.py

### CODE REVIEW

**Comment by WoosukKwon** [comment]
cc @njhill Could you please take a look?

File: vllm/v1/worker/gpu_input_batch.py [hunk 1]
```diff
@@ -195,6 +195,10 @@ def __init__(
         self.logit_bias: List[Optional[Dict[int,
                                             float]]] = [None] * max_num_reqs
 
+        self.has_allowed_token_ids: List[bool] = [False] * max_num_reqs
+        self.allowed_token_ids_mask: torch.Tensor = torch.zeros(
+            max_num_reqs, self.vocab_size, dtype=torch.bool, device=device)
```
Comment (njhill) [inline vllm/v1/worker/gpu_input_batch.py:200]: This could actually get quite large (e.g. 128 MB), we may want to consider a different (pre)allocation strategy.

Also, consistent with the other tensors, we should keep this in pinned CPU memory and only copy to GPU in an async manner when the batch makeup changes (and has_allowed_token_ids is True)

File: vllm/v1/sample/sampler.py [hunk 2]
```diff
@@ -227,3 +237,17 @@ def apply_logits_bias(
...
+        logits: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+    ) -> torch.Tensor:
+        # One idea is implement this as a PyTorch C++ op, and we may
+        # even optimize the logit_bias layout.
+        for i, allowed_token_ids in enumerate(
+                sampling_metadata.has_allowed_token_ids):
+            if allowed_token_ids:
+                logits[i].masked_fill_(
+                    sampling_metadata.allowed_token_ids_mask[i], float("-inf"))
```
Comment (njhill) [inline vllm/v1/sample/sampler.py:252]: Can we do a single operation here rather than a loop (make `has_allowed_token_ids` a single bool rather than list).

File: tests/v1/sample/test_sampler.py [hunk 3]
```diff
@@ -412,3 +441,44 @@ def test_sampler_logit_bias(device: str, batch_size: int, bias_value: float):
                                                                  1e-2)
             else:
                 assert logits_for_req[token_id] == pytest.approx(1e-2)
+
+
+@pytest.mark.parametrize("device", CUDA_DEVICES)
+@pytest.mark.parametrize("batch_size", [1, 2, 32])
+@pytest.mark.parametrize("num_allowed_token_ids", [0, 1, 2])
+def test_sampler_allowed_token_ids(device: str, batch_size: int,
```
Comment (njhill) [inline tests/v1/sample/test_sampler.py:449]: Thank you for adding this test!!

**Review by njhill — COMMENTED** [review]
Thanks @houseroad!

I have a few comments, mainly to keep the tensor manipulation on cpu and apply mask to the batch rather than individual rows.

Also you might want to take a look at https://github.com/vllm-project/vllm/pull/13360, I'm hoping to get that into shape soon and have all the sampling params like this encapsulated as vectorized logits processors.

File: vllm/v1/sample/sampler.py [hunk 4]
```diff
@@ -47,17 +47,22 @@ def forward(
...
         # Use float32 for the logits.
         logits = logits.to(torch.float32)
+        # Apply allowed token ids.
+        logits = self.apply_allowed_token_ids(logits, sampling_metadata)
         # Apply logits bias.
         logits = self.apply_logits_bias(logits, sampling_metadata)
         # Apply penalties (e.g., min_tokens, freq_penalties).
         logits = self.apply_penalties(logits, sampling_metadata)
+        # Apply temperature.
+        logits = self.apply_temperature(logits, sampling_metadata.temperature)
```
[the remark this answers is not in the document]
Comment (houseroad) [reply vllm/v1/sample/sampler.py]: Good catch

File: vllm/v1/sample/sampler.py (hunk 2 above)
Comment (houseroad) [reply vllm/v1/sample/sampler.py]: Yeah, just did everything in one masked_fill_

File: vllm/v1/sample/metadata.py [hunk 5]
```diff
@@ -37,3 +37,11 @@ class SamplingMetadata:
...
 
     logit_bias: List[Optional[Dict[int, float]]]
+
+    # These two parameters are for allowed_token_ids.
+    # `no_allowed_token_ids`` is a bool to indicate whether we have
+    # allowed_token_ids.
+    # `allowed_token_ids_mask` is a 2D bool tensor of shape (max batch size,
+    # vocab size).
+    no_allowed_token_ids: bool
+    allowed_token_ids_mask: Optional[torch.Tensor]
```
Comment (njhill) [inline vllm/v1/sample/metadata.py:47]: Actually I don't think we need the bool since that's what `None` means? (similar to how `top_k`, `top_p`, etc. are now).

File: vllm/v1/sample/sampler.py [hunk 6]
```diff
@@ -181,14 +183,17 @@ def apply_penalties(
...
             logits = apply_all_penalties(
-                logits, sampling_metadata.prompt_token_ids,
+                logits,
+                sampling_metadata.prompt_token_ids,
                 sampling_metadata.presence_penalties,
                 sampling_metadata.frequency_penalties,
                 sampling_metadata.repetition_penalties,
-                sampling_metadata.output_token_ids)
+                sampling_metadata.output_token_ids,
+            )
```
Comment (njhill) [inline vllm/v1/sample/sampler.py:196]: Same general comment, it's best not to include unrelated formatting changes, looks like there are a couple above and some in the other file.

File: vllm/v1/sample/sampler.py [hunk 7]
```diff
@@ -226,3 +231,15 @@ def apply_logits_bias(
...
         return logits
+
+    def apply_allowed_token_ids(
+        self,
+        logits: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+    ) -> torch.Tensor:
+        # One idea is implement this as a PyTorch C++ op, and we may
+        # even optimize the logit_bias layout.
+        if not sampling_metadata.no_allowed_token_ids:
```
Comment (njhill) [inline vllm/v1/sample/sampler.py:242]: per other comment, this can be
```suggestion
        if sampling_metadata.allowed_token_ids_mask is not None:
```

File: vllm/v1/worker/gpu_input_batch.py [hunk 8]
```diff
@@ -192,6 +192,9 @@ def __init__(
 
         self.logit_bias: List[Optional[Dict[int,
                                             float]]] = [None] * max_num_reqs
+        self.has_allowed_token_ids: List[bool] = [False] * max_num_reqs
```
Comment (njhill) [inline vllm/v1/worker/gpu_input_batch.py:195]: Similar to the other params, better for this to be a set of request indices. Then `no_allowed_token_ids` can just check whether the set is empty without having to iterate, and in `remove_request`, if the index was not already in the set, no need to clear that row of the mask.

File: vllm/v1/worker/gpu_input_batch.py [hunk 9]
```diff
@@ -436,6 +471,13 @@ def _make_sampling_metadata(self) -> SamplingMetadata:
         else:
             prompt_token_ids = None
 
+        allowed_token_ids_mask: Optional[torch.Tensor] = None
+        if not self.no_allowed_token_ids and \
+                self.allowed_token_ids_mask is not None:
```
Comment (njhill) [inline vllm/v1/worker/gpu_input_batch.py:476]: I think no need for the second condition here since it will always have been allocated if the first condition is true.

**Review by njhill — COMMENTED** [review]
Thanks @lu-wang-dl for the updates, I think it's close, just a few more things.

File: vllm/v1/sample/metadata.py (hunk 5 above)
Comment (houseroad) [reply vllm/v1/sample/metadata.py:47]: There are cases, if allowed_token_ids_mask got initialized, it will not go back again. So I am still trying to keep the bool to indicate whether we indeed need to apply allowed_token_ids.

File: vllm/v1/sample/sampler.py (hunk 6 above)
Comment (houseroad) [reply vllm/v1/sample/sampler.py:196]: yeah, it's done by my VSCode, sorry.

I will be more cautious next time.

File: vllm/v1/sample/metadata.py (hunk 5 above)
Comment (njhill) [reply vllm/v1/sample/metadata.py:47]: Right but the bool isn't needed for that in the metadata. You can set it dependent on the `no_allowed_token_ids` property (whether the set is empty) ... same as the top_p etc. fields in here if you look at where they're set. That's independent of the lazy loading.

File: vllm/v1/sample/sampler.py (hunk 7 above)
Comment (houseroad) [reply vllm/v1/sample/sampler.py:242]: allowed_token_ids_mask can be initialized, but no request leverage it anymore. Then we still applying masked_fill_, which could be unnecessary.

File: vllm/v1/worker/gpu_input_batch.py (hunk 9 above)
Comment (houseroad) [reply vllm/v1/worker/gpu_input_batch.py:476]: this is purely for type refinement. Otherwise, mypy will yell at me. I think it cannot tell the type of self. allowed_token_ids_mask[:num_req] if it doesn't know whether allowed_token_ids_mask is none or not.

Comment (njhill) [reply vllm/v1/worker/gpu_input_batch.py:476]: It's better in that case to `assert allowed_token_ids_mask is not None` within the `if`

File: vllm/v1/sample/sampler.py (hunk 7 above)
Comment (njhill) [reply vllm/v1/sample/sampler.py:242]: Not if we make `None` in the sampling metadata be equivalent to the `no_allowed_token_ids` flag (and so can get rid of that flag per the other [comment](https://github.com/vllm-project/vllm/pull/13210#discussion_r1962470972)). Note that it could still be set to None in the sampling metadata even though it's been initialized and non-None in the input batch.

Comment (houseroad) [reply vllm/v1/sample/sampler.py]: good point, addressed

File: vllm/v1/worker/gpu_input_batch.py (hunk 8 above)
Comment (houseroad) [reply vllm/v1/worker/gpu_input_batch.py]: yeah, switched to a set

File: vllm/v1/worker/gpu_input_batch.py (hunk 9 above)
Comment (houseroad) [reply vllm/v1/worker/gpu_input_batch.py]: yup, good idea.

File: vllm/v1/sample/sampler.py (hunk 6 above)
Comment (houseroad) [reply vllm/v1/sample/sampler.py:196]: okay, confirmed, actually it's done by the yapf in the pre-commit hooks. It forced me to do such change...

File: vllm/v1/worker/gpu_input_batch.py [hunk 10]
```diff
@@ -285,6 +288,28 @@ def add_request(
         if sampling_params.logit_bias is not None:
             self.logit_bias[req_index] = sampling_params.logit_bias
 
+        if sampling_params.allowed_token_ids:
+            # NOTE(houseroad): put the check here since no vocab_size info
+            # available in vllm/sampling_params.py
+            if not all(0 <= tid < self.vocab_size
+                       for tid in sampling_params.allowed_token_ids):
+                raise ValueError(
+                    "allowed_token_ids contains out-of-vocab token id")
```
Comment (njhill) [inline vllm/v1/worker/gpu_input_batch.py:297]: I think this will be a problem because it will cause the entire batch to fail.

Could you move the validation to [here](https://github.com/vllm-project/vllm/blob/main/vllm/v1/engine/processor.py#L103)?

**Review by njhill — COMMENTED** [review]
Thanks @houseroad looks great now, just one remaining thing.

File: vllm/v1/worker/gpu_input_batch.py (hunk 10 above)
Comment (houseroad) [reply vllm/v1/worker/gpu_input_batch.py]: Yeah, good point. Moved.

**Review by njhill — APPROVED** [review]
Thanks again @houseroad!
